File: histogram/Python/motion_detector.py
import cv2
import numpy as np
import vlc


def equalized_bgr_frame(frame: np.ndarray) -> list:
    equalized_channels = list()

    for channel in cv2.split(frame):
        equalized_channels.append(cv2.equalizeHist(channel))

    return equalized_channels


def stack_to_show(frame: np.ndarray, channels: list) -> np.ndarray:
    hs_stacking = list()
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)

    for channel in channels:
        hs_stacking.append(np.hstack((gray, channel)))

    return np.vstack(tuple(hs_stacking))

    # exit(1)


def motion_detector(old_equalized_channels: list, new_equalized_channels: list) -> bool:
    all_dis = list()
    if old_equalized_channels:
        for old_channel, new_channel in zip(old_equalized_channels, new_equalized_channels):
            dist = np.linalg.norm((old_channel.flatten() / 255) - (new_channel.flatten() / 255))
            all_dis.append(dist)

        if np.max(all_dis) > 50:
            return True
        else:
            return False

    else:
        return False


def main():
    cam = cv2.VideoCapture(0)

    window_name = "frame"
    cv2.namedWindow(window_name, cv2.WINDOW_NORMAL)

    player = vlc.MediaPlayer("music/babyshark.mp3")

    old_equalized_channels = list()

    if not cam.isOpened():
        print("não achou a camera !!")
        exit(1)

    while "e" != chr(cv2.waitKey(1) & 255):
        ret, frame = cam.read()
        if ret:
            new_equalized_channels = equalized_bgr_frame(frame)
            if motion_detector(old_equalized_channels, new_equalized_channels):
                if not player.get_state():
                    player.play()

            old_equalized_channels = new_equalized_channels
            cv2.imshow(window_name, frame)
    pass


if __name__ == '__main__':
    main()

File: histogram/C++/motionDetector.cpp
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/* 
 * File:   motionDetector.cpp
 * Author: samuel
 *
 * Created on August 25, 2019, 3:55 PM
 */

#include <iostream>
#include <opencv2/opencv.hpp>
#include <cstdlib>
#include <vlc/vlc.h>

using namespace std;


bool motionDetector(std::vector<cv::Mat> *&oldChannels, cv::Mat &frame);

double getDist(cv::Mat &oldChannel, cv::Mat &newChannel);

int main(int argc, char **argv) {
    cv::VideoCapture cam = cv::VideoCapture(0);
    std::string windowName = "Cam";
    std::string exitKey = "e";
    cv::Mat frame;
    std::vector<cv::Mat> *oldChannels = NULL;
    libvlc_instance_t *inst;
    libvlc_media_player_t *mediaPlayer;
    libvlc_media_t *media;
    inst = libvlc_new(0, NULL);

    cv::namedWindow(windowName, cv::WINDOW_AUTOSIZE);

    if (not cam.isOpened()) {
        std::cout << "Camera não encontrada" << std::endl;
        std::exit(1);
    }

    if (argc < 2) {
        std::cout << " passe como parametro o caminho da musica!" << std::endl;
        std::exit(-1);
    }

    media = libvlc_media_new_path(inst, argv[1]);
    mediaPlayer = libvlc_media_player_new_from_media(media);

    // não precisa mais da midia
    libvlc_media_release(media);


    while (exitKey != std::string(1, (char) (cv::waitKey(1) & 255))) {
        if (cam.read(frame)) {
            cv::imshow(windowName, frame);

            if (motionDetector(oldChannels, frame))
                if (not libvlc_media_player_is_playing(mediaPlayer))
                    libvlc_media_player_play(mediaPlayer);
        }

    }

    return 0;
}


bool motionDetector(std::vector<cv::Mat> *&oldChannels, cv::Mat &frame) {
    auto equalizedChannels = new std::vector<cv::Mat>(3);
    bool detected = false;
    double bodyMoved = 0.2;
    cv::Mat equalizedHist;
    cv::split(frame, *equalizedChannels);

    if (oldChannels) {
        for (int i = 0; i < equalizedChannels->size(); i++) {
            cv::equalizeHist(equalizedChannels->at(i), equalizedHist);

            if (getDist(equalizedHist, oldChannels->at(i)) > bodyMoved)
                detected = true;
        }
    } else
        for (auto &equalizedChannel : *equalizedChannels)
            cv::equalizeHist(equalizedChannel, equalizedChannel);

    oldChannels = equalizedChannels;

    return detected;
}

double getDist(cv::Mat &oldChannel, cv::Mat &newChannel) {
    int cv_32f = 5;
    cv::Mat oldChannelFloat;
    cv::Mat newChannelFloat;

    oldChannel.convertTo(oldChannelFloat, cv_32f);
    newChannel.convertTo(newChannelFloat, cv_32f);
    cv::normalize(oldChannelFloat, oldChannelFloat);
    cv::normalize(newChannelFloat, newChannelFloat);

    return cv::norm(oldChannelFloat, newChannelFloat);

}